pgxn/neon_walredo/walredoproc.c

```diff
@@ -226,7 +243,146 @@ WalRedoMain(int argc, char *argv[])
 	InitializeWalConsistencyChecking();
 #endif
 
-	CreateSharedMemoryAndSemaphores();
+	/*
+	 * Initialize dummy shmem.
+	 *
+	 * This code follows CreateSharedMemoryAndSemaphores() but manually sets up
+	 * the shmem header and skips few initializations that are not needed for
+	 * WAL redo.
+	 */
+	PGShmemHeader *shim = NULL;
+	{
+		PGShmemHeader *hdr;
+		Size size = 10*1048576; // 20 mb. Default nbuffers are 1024, so 8mb shmem
+
+		hdr = (PGShmemHeader *) malloc(size);
+		if (!hdr)
+			ereport(ERROR,
+					(errcode(ERRCODE_OUT_OF_MEMORY),
+					errmsg("[neon-wal-redo] can't allocate (pseudo-) shared memory")));
+
+		hdr->creatorPID = getpid();
+		hdr->magic = PGShmemMagic;
+		hdr->dsm_control = 0;
+		hdr->device = 42; // not relevant for non-shared memory
+		hdr->inode = 424242; // not relevant for non-shared memory
+		hdr->totalsize = size;
+		hdr->freeoffset = MAXALIGN(sizeof(PGShmemHeader));
+
+		shim = hdr;
+		UsedShmemSegAddr = hdr;
+		UsedShmemSegID = (unsigned long) 42;
```

Comment:
This part replaces PGSharedMemoryCreate. Let's add a comment to make that explicit, or put it to a separate `FakeSharedMemoryCreate`  function, for clarity